### PR TITLE
Add FasterPath#dirname method and write some tests on it

### PR DESCRIPTION
**Pull request by gogainda**
I've implemented wrapper method for dirname and ported some tests from rubinius project.

Seems that some tests are failing(marked them in the code with #failing comment)

@danielpclark  Speaking about tests, should we fix the rust implementation or just fixt failing tests?

### CODE REVIEW

**Comment by gogainda** [comment]
Task is done, but turns out that my implementation of dirname in rust is slowe than ruby's:
```
 PASS (5.81s) :: bench_ruby_chop_basename
bench_rust_dirname       0.030489        0.030671        0.031002        0.030689        0.031651
 PASS (0.21s) :: bench_rust_dirname
bench_ruby_dirname       0.018665        0.018959        0.017234        0.018418        0.017791
```

**Comment by gogainda** [comment]
@danielpclark

**Comment by danielpclark** [comment]
I don't know if this helps but here's the C source code for this method https://github.com/ruby/ruby/blob/32674b167bddc0d737c38f84722986b0f228b44b/file.c#L4118-L4158 .  I'm unfamiliar with the code base so it would take some time for me to fully grok.

**Comment by danielpclark** [comment]
We can add this in if you update it with the `WITH_REGRESSION` env flag. See: https://github.com/danielpclark/faster_path/wiki/Policies-on-Reimplementing-C-Code

**Comment by gogainda** [comment]
Ok, will update in couple of days

File: lib/faster_path/optional/monkeypatches.rb [hunk 1]
```diff
@@ -3,7 +3,11 @@ def self.sledgehammer_everything!
     ::File.class_eval do
       def basename(pth)
         FasterPath.basename(pth)
-      end 
+      end
+
+      def dirname(pth)
```
Comment (glebm) [inline lib/faster_path/optional/monkeypatches.rb]: Should be `self.dirname`. As it is, this doesn't override the existing method but adds an instance method instead.

**Comment by danielpclark** [comment]
I've added a new benchmark system called pbench.  You can run `rake pbench` to see how much the method performance has improved.  You will need to have benchmark code added to `test/pbench/suite.rb` to see it.  This is brand new on the master branch.

**Comment by gogainda** [comment]
@danielpclark review pls
```
32-bit ruby 2.2.0p0 (2014-12-25 revision 49005) [i686-linux]
32-bit rustc 1.11.0-nightly (5c2a5d449 2016-06-11)
--------------------------------------------------------------------------------
Performance change for allocate, instead of new, is 74.1%
Performance change for absolute? is 91.1%
Performance change for add_trailing_separator is 67.6%
Performance change for basename is -18.2%
Performance change for chop_basename is 22.5%
Performance change for directory? is -4.2%
Performance change for extname is 37.3%
Performance change for dirname is -63.7%
Performance change for has_trailing_separator? is 60.6%
Performance change for blank? (verses strip.empty?) is -59.4%
Performance change for relative? is 90.7%
```

**Comment by gogainda** [comment]
@danielpclark ready for merge

**Comment by danielpclark** [comment]
Good work.

Change `use std::path::{Path,MAIN_SEPARATOR};` to `use std::path::MAIN_SEPARATOR;` and change the unused `usize` to an underscore in the areas below.

```
src/dirname.rs:1:17: 1:21 warning: unused import, #[warn(unused_imports)] on by default
src/dirname.rs:1 use std::path::{Path,MAIN_SEPARATOR};
                                 ^~~~
src/dirname.rs:25:18: 25:23 warning: unused variable: `usize`, #[warn(unused_variables)] on by default
src/dirname.rs:25         .find(|&(usize, char)| char != MAIN_SEPARATOR);
                                   ^~~~~
src/dirname.rs:34:18: 34:23 warning: unused variable: `usize`, #[warn(unused_variables)] on by default
src/dirname.rs:34         .find(|&(usize, char)| char == MAIN_SEPARATOR);
                                   ^~~~~
src/dirname.rs:57:22: 57:27 warning: unused variable: `usize`, #[warn(unused_variables)] on by default
src/dirname.rs:57             .find(|&(usize, char)| char != MAIN_SEPARATOR);
                                       ^~~~~
```

After that copy the benchmarks you made into the pbench suit in `test/pbench/pbench_suite.rb` and you're done!

**Pull request merged**